File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if (len(s) <= 1):
            return len(s)
        
        seenChars = {};
        longest = 0;
        left = 0;
        right = 0;
        
        while (left < len(s) and right < len(s)):
            if (s[right] in seenChars):
                left = max(left, seenChars[s[right]] + 1)
            seenChars[s[right]] = right
            longest = max(longest, right - left + 1)
            right += 1
        
        return longest

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (s.length <= 1) {
        return s.length;
    }
    
    let maxLength = 0;    
    let currString = {}
    let right = 0;
    let left = 0;
    
    while (right < s.length && left < s.length) {
        if (s[right] in currString) {
            left = Math.max(left, currString[s[right]] + 1);
        }
        currString[s[right]] = right;
        maxLength = Math.max(maxLength, (right - left + 1));
        right++;
    }

    return maxLength;
};